keras_cv/applications/object_detection/retina_net/retina_net.py

```diff
@@ -0,0 +1,325 @@
+# Copyright 2022 The KerasCV Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import numpy as np
+import tensorflow as tf
+from tensorflow import keras
+
+from keras_cv import bounding_box
+from keras_cv.applications.object_detection.retina_net.__internal__ import (
+    layers as layers_lib,
+)
+from keras_cv.applications.object_detection.retina_net.__internal__ import (
+    utils as utils_lib,
+)
+
+
+# TODO(lukewood): update docstring to include documentation on creating a custom label
+# decoder/etc.
+# TODO(lukewood): link to keras.io guide on creating custom backbone and FPN.
+class RetinaNet(keras.Model):
+    """A Keras model implementing the RetinaNet architecture.
+
+    Implements the RetinaNet architecture for object detection.  The constructor
+    requires `classes`, `bounding_box_format` and a `backbone`.  Optionally, a
+    custom label encoder, feature pyramid network, and prediction decoder may all be
+    provided.
+
+    Usage:
+    ```python
+    retina_net = keras_cv.applications.RetinaNet(
+        classes=20,
+        bounding_box_format="xywh",
+        backbone="resnet50",
+        backbone_weights="imagenet",
+        include_rescaling=True,
+    )
+    ```
+
+    Args:
+        classes: the number of classes in your dataset excluding the background
+            class.  Classes should be represented by integers in the range
+            [0, classes).
+        bounding_box_format: The format of bounding boxes of input dataset. Refer
+            [to the keras.io docs](https://keras.io/api/keras_cv/bounding_box/formats/)
+            for more details on supported bounding box formats.
+        backbone: Either `"resnet50"` or a custom backbone model.
```

Comment:
Just a thought: should we have something in KerasCV to automatically turn any KerasCV.model that is an image classifier into a backbone?

Reply:
Unfortunately we can't do this.  too many details, i.e. what the best layers to take outputs from are.